File: java/p2/src/p2/types/element/ScanJoin.java
package p2.types.element;

import p2.types.basic.SimpleTupleSet;
import p2.types.basic.Tuple;
import p2.types.basic.TupleSet;
import p2.types.table.Table;

public class ScanJoin extends Operator {
	
	/* The table to be scanned. */
	private Table table;
	
	private Integer[] outerKey;
	
	private Integer[] innerKey;

	public ScanJoin(String id, String name, Table table, Integer[] outerKey, Integer[] innerKey) {
		super(id, name);
		assert(outerKey.length == innerKey.length);
		this.table = table;
		this.outerKey = outerKey;
		this.innerKey = innerKey;
	}
	
	/**
	 * Determine the join result of the outerSet and the inner table.
	 */
	public TupleSet simple_action(TupleSet outerSet) {
		TupleSet result = new SimpleTupleSet();
		for (Tuple inner : table) {
			for (Tuple outer : outerSet) {
				result.add(Tuple.join(outer, inner, innerKey));
			}
		}
		return result;
	}

}

File: lincoln/ruby/lib/types/operator/scan_join.rb
require 'lib/types/operator/join'
require 'lib/types/table/table'

class ScanJoin < Join 
	def initialize(predicate, input)
		super(predicate, input)
		@table = Table::table(predicate.name)
	end
	
	def to_s
		return "NEST LOOP JOIN: PREDICATE[" + @predicate  + "]"
	end
	
	def evaluate(tuples) #TupleSet
		result = TupleSet.new("tmp", nil)
		innerTuples = @table.tuples
		tuples.each do |outer| 
		  innerTuples.each do |inner|
				inner.schema = @predicate.schema.clone
				
				if validate(outer, inner)
					join = outer.join(inner)
					if (!join.nil?) 
						result << join
					end
				end
			end
		end
		return result
	end

end
